### ISSUE TITLE
cupy-6.0.0 fails to import on Python 3.8.0b1 for Windows

### ISSUE THREAD

**Issue by cgohlke** [cupy/cupy#2243]
Just a heads-up: using CuPy 6.0.0 built from source with Cython-0.29.10 and CUDA 10.1 on Python 3.8.0b1, Windows 10, cupy fails to import (at least on my system, traceback below). This is due to a [change of the DLL resolution](https://docs.python.org/3.8/whatsnew/3.8.html#bpo-36085-whatsnew) on that platform. A quick workaround if `CUDA_PATH` is defined:

```patch
diff --git a/cupy/__init__.py b/cupy/__init__.py
index 006e60cc3..8d4973193 100644
--- a/cupy/__init__.py
+++ b/cupy/__init__.py
@@ -6,6 +6,11 @@ import six

 from cupy import _version

+try:
+    import os
+    os.add_dll_directory(os.path.join(os.environ['CUDA_PATH'], 'bin'))
+except Exception:
+    pass

 try:
     from cupy import core  # NOQA
```

```python-traceback
> python.exe -c"import cupy"
Traceback (most recent call last):
  File "X:\Python38\lib\site-packages\cupy\__init__.py", line 11, in <module>
    from cupy import core  # NOQA
  File "X:\Python38\lib\site-packages\cupy\core\__init__.py", line 1, in <module>
    from cupy.core import core  # NOQA
  File "cupy\core\core.pyx", line 1, in init cupy.core.core
  File "X:\Python38\lib\site-packages\cupy\cuda\__init__.py", line 4, in <module>
    from cupy.cuda import compiler  # NOQA
  File "X:\Python38\lib\site-packages\cupy\cuda\compiler.py", line 11, in <module>
    from cupy.cuda import device
  File "cupy\cuda\runtime.pxd", line 8, in init cupy.cuda.device
ImportError: DLL load failed: The specified module could not be found.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "X:\Python38\lib\site-packages\cupy\__init__.py", line 32, in <module>
    six.reraise(ImportError, ImportError(msg), exc_info[2])
  File "X:\Python38\lib\site-packages\six.py", line 692, in reraise
    raise value.with_traceback(tb)
  File "X:\Python38\lib\site-packages\cupy\__init__.py", line 11, in <module>
    from cupy import core  # NOQA
  File "X:\Python38\lib\site-packages\cupy\core\__init__.py", line 1, in <module>
    from cupy.core import core  # NOQA
  File "cupy\core\core.pyx", line 1, in init cupy.core.core
  File "X:\Python38\lib\site-packages\cupy\cuda\__init__.py", line 4, in <module>
    from cupy.cuda import compiler  # NOQA
  File "X:\Python38\lib\site-packages\cupy\cuda\compiler.py", line 11, in <module>
    from cupy.cuda import device
  File "cupy\cuda\runtime.pxd", line 8, in init cupy.cuda.device
ImportError: CuPy is not correctly installed.

If you are using wheel distribution (cupy-cudaXX), make sure that the version of CuPy you installed matches with the version of CUDA on your host.
Also, confirm that only one CuPy package is installed:
  $ pip freeze

If you are building CuPy from source, please check your environment, uninstall CuPy and reinstall it with:
  $ pip install cupy --no-cache-dir -vvvv

Check the Installation Guide for details:
  https://docs-cupy.chainer.org/en/latest/install.html

original error: DLL load failed: The specified module could not be found.
```

**Comment by C9KenData** [comment]
i think i'm running into the same problem
but i couldn't quite understand what the solution means. 

i checked environmental variables, my CUDA_PATH is defined (default).
what do i do with the script?

**Comment by cgohlke** [comment]
> i think i'm running into the same problem
but i couldn't quite understand what the solution means.

Add the following lines to your code before importing cupy or add them to `cupy/__init__.py`:
```Python
try:
    import os
    os.add_dll_directory(os.path.join(os.environ['CUDA_PATH'], 'bin'))
except Exception:
    pass
```

**Comment by C9KenData** [comment]
I ran this 

```
try:
    import os
    os.add_dll_directory(os.path.join(os.environ['CUDA_PATH'], 'bin'))
except Exception:
    pass

import cupy

```
and got this: 

'''
 runfile('C:/Users/New User/Documents/shopify/tester.py', wdir='C:/Users/New User/Documents/shopify')
Traceback (most recent call last):

  File "<ipython-input-4-e60f7b3f5e8e>", line 1, in <module>
    runfile('C:/Users/New User/Documents/shopify/tester.py', wdir='C:/Users/New User/Documents/shopify')

  File "C:\ProgramData\Anaconda3\lib\site-packages\spyder_kernels\customize\spydercustomize.py", line 827, in runfile
    execfile(filename, namespace)

  File "C:\ProgramData\Anaconda3\lib\site-packages\spyder_kernels\customize\spydercustomize.py", line 110, in execfile
    exec(compile(f.read(), filename, 'exec'), namespace)

  File "C:/Users/New User/Documents/shopify/tester.py", line 14, in <module>
    import cupy

  File "C:\ProgramData\Anaconda3\lib\site-packages\cupy\__init__.py", line 11, in <module>
    from cupy import core  # NOQA

  File "C:\ProgramData\Anaconda3\lib\site-packages\cupy\core\__init__.py", line 1, in <module>
    from cupy.core import core  # NOQA

  File "cupy\core\carray.pxi", line 50, in init cupy.core.core

AttributeError: type object 'cupy.core.core.Indexer' has no attribute '__reduce_cython__'
'''

**Comment by C9KenData** [comment]
os doesn't have a add_dll_director function.... 
and my computer can't find CUDA_PATH despite me adding it onto environmental variables. =/

**Comment by cgohlke** [comment]
> os doesn't have a add_dll_director function

This issue and workaround are specific to using official Python 3.8 and the NVIDIA CUDA Toolkit on Windows. They don't apply other Python versions or Conda distributions...

**Comment by C9KenData** [comment]
i see. i assume i can manually add the dll path ? 
if i know where to find it ...

**Comment by C9KenData** [comment]
went to environmental variable, added the bin path and i can import cupy now

**Comment by kmaehashi** [comment]
Fixed in v6.6.0/v7.0.0.

**Comment by dome272** [comment]
Im getting the same issue as @cgohlke 
Im installed cuda 10.2 and installed `pip install cupy-cuda102 --no-cache-dir -vvvv` (Initially I installed just `pip install cupy-cuda102` but it didnt work either)
Furthermore Im using Python 3.6.5
Error: 
`
>>> import cupy
Traceback (most recent call last):
  File "C:\Users\youre\AppData\Local\Programs\Python\Python36\lib\site-packages\cupy\__init__.py", line 21, in <module>
    from cupy import core  # NOQA
  File "C:\Users\youre\AppData\Local\Programs\Python\Python36\lib\site-packages\cupy\core\__init__.py", line 1, in <module>
    from cupy.core import core  # NOQA
  File "cupy\core\core.pyx", line 1, in init cupy.core.core
  File "C:\Users\youre\AppData\Local\Programs\Python\Python36\lib\site-packages\cupy\cuda\__init__.py", line 5, in <module>
    from cupy.cuda import compiler  # NOQA
  File "C:\Users\youre\AppData\Local\Programs\Python\Python36\lib\site-packages\cupy\cuda\compiler.py", line 13, in <module>
    from cupy.cuda import device
  File "cupy\cuda\runtime.pxd", line 8, in init cupy.cuda.device
ImportError: DLL load failed: The specified module could not be found.
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\youre\AppData\Local\Programs\Python\Python36\lib\site-packages\cupy\__init__.py", line 42, in <module>
    six.reraise(ImportError, ImportError(msg), exc_info[2])
  File "C:\Users\youre\AppData\Local\Programs\Python\Python36\lib\site-packages\six.py", line 702, in reraise
    raise value.with_traceback(tb)
  File "C:\Users\youre\AppData\Local\Programs\Python\Python36\lib\site-packages\cupy\__init__.py", line 21, in <module>
    from cupy import core  # NOQA
  File "C:\Users\youre\AppData\Local\Programs\Python\Python36\lib\site-packages\cupy\core\__init__.py", line 1, in <module>
    from cupy.core import core  # NOQA
  File "cupy\core\core.pyx", line 1, in init cupy.core.core
  File "C:\Users\youre\AppData\Local\Programs\Python\Python36\lib\site-packages\cupy\cuda\__init__.py", line 5, in <module>
    from cupy.cuda import compiler  # NOQA
  File "C:\Users\youre\AppData\Local\Programs\Python\Python36\lib\site-packages\cupy\cuda\compiler.py", line 13, in <module>
    from cupy.cuda import device
  File "cupy\cuda\runtime.pxd", line 8, in init cupy.cuda.device
ImportError: CuPy is not correctly installed.
If you are using wheel distribution (cupy-cudaXX), make sure that the version of CuPy you installed matches with the version of CUDA on your host.
Also, confirm that only one CuPy package is installed:
  $ pip freeze
If you are building CuPy from source, please check your environment, uninstall CuPy and reinstall it with:
  $ pip install cupy --no-cache-dir -vvvv
Check the Installation Guide for details:
  https://docs-cupy.chainer.org/en/latest/install.html
original error: DLL load failed: The specified module could not be found.
`

As suggested maybe it has something to do with the env variables but Im not sure what I should edit there. The CUDA_PATH seems to be set `C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v10.2`

Any help would be appreciated...
